src/ripple/consensus/Validations.h

```diff
@@ -518,211 +618,281 @@ class Validations
+            });
 
-        current(
-            stalePolicy_.now(),
-            // The number of validations does not correspond to the number of
-            // distinct ledgerIDs so we do not call reserve on ret.
-            [](std::size_t) {},
-            [this,
-             &cutoffBefore,
-             &currentLedger,
-             &valCurrentLedger,
-             &valPriorLedger,
-             &priorLedger,
-             &ret](NodeKey const&, ValidationAndPrevID const& vp) {
-                Validation const& v = vp.val;
-                LedgerID const& prevLedgerID = vp.prevLedgerID;
-                if (!v.trusted())
-                    return;
-
-                SeqType const seq = v.seq();
-                if ((seq == SeqType{0}) || (seq >= cutoffBefore))
-                {
-                    // contains a live record
-                    bool countPreferred =
-                        valCurrentLedger && (v.ledgerID() == currentLedger);
-
-                    if (!countPreferred &&  // allow up to one ledger slip in
-                                            // either direction
-                        ((valCurrentLedger &&
-                          (prevLedgerID == currentLedger)) ||
-                         (valPriorLedger && (v.ledgerID() == priorLedger))))
-                    {
-                        countPreferred = true;
-                        JLOG(this->j_.trace()) << "Counting for " << currentLedger
-                                         << " not " << v.ledgerID();
-                    }
-
-                    if (countPreferred)
-                        ret[currentLedger]++;
-                    else
-                        ret[v.ledgerID()]++;
-                }
+        // No trusted validations to determine branch
+        if (preferredSeq == Seq{0})
+            return std::make_pair(preferredSeq, preferredID);
+
+        Seq currSeq = currLedger.seq();
+        ID currID = currLedger.id();
+
+        // If we are the parent of the preferred ledger, stick with our
+        // current ledger since we might be about to generate it
+        if (preferredSeq == currSeq + Seq{1})
+        {
+            for (auto const& it : lastLedger_)
+            {
+                Ledger const& ledger = it.second;
+                if (ledger.seq() == preferredSeq &&
+                    ledger.id() == preferredID && ledger[currSeq] == currID)
+                    return std::make_pair(currSeq, currID);
+            }
+        }
+
+        // A ledger ahead of us is preferred regardless of whether it is
+        // a descendent of our working ledger or it is on a different chain
+        if (preferredSeq > currSeq)
+            return std::make_pair(preferredSeq, preferredID);
+
+        // Only switch to earlier sequence numbers if it is a different chain.
+        if (currLedger[preferredSeq] != preferredID)
+            return std::make_pair(preferredSeq, preferredID);
```

Comment:
It looks like this can cause us to switch from our current ledger to a preferred ledger with the same seq but different id. If that's correct, maybe mention that case in the comment above.